src/Microsoft.Identity.Client/Mats/Mats.cs

```diff
@@ -154,6 +154,39 @@ public void EndAction(IActionHandle action, AuthOutcome outcome, ErrorSource err
             _actionStore.EndAdalAction((AdalAction)action, outcome, errorSource, error, errorDescription);
             _uploader.Upload(GetEventsForUpload());
         }
+
+        public void EndAction(IActionHandle action, Exception ex)
+        {
+            AuthOutcome outcome = AuthOutcome.Failed;
+            ErrorSource errorSource = ErrorSource.AuthSdk;
+            string error = ex.Message;
+            string errorDescription = ex.Message;
+
+            switch (ex)
+            {
+            case MsalUiRequiredException uiEx:
+                errorSource = ErrorSource.Service;
+                error = uiEx.ErrorCode;
+                break;
+            case MsalServiceException svcEx:
+                errorSource = ErrorSource.Service;
+                error = svcEx.ErrorCode;
+                break;
+            case MsalClientException cliEx:
+                errorSource = ErrorSource.Client;
+                error = cliEx.ErrorCode;
+                break;
+            case MsalException msalEx:
+                errorSource = ErrorSource.AuthSdk;
+                error = msalEx.ErrorCode;
+                break;
+            default:
+                errorSource = ErrorSource.AuthSdk;
```

Comment:
It's not clear from looking at the enum that the error source "auth sdk" actually means unhandled exception in MSAL.

Reply:
I need to get clarity from MATS on the specificities of when to use ErrorSource.Client versus ErrorSource.AuthSdk.  I agree they're very similar.

---
In reply to: [264336430](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/955#discussion_r264336430) [](ancestors = 264336430)